### ISSUE TITLE
[DF] select * limit 5 seems does a full scan

### ISSUE THREAD

**Issue by randerzander** [dask-contrib/dask-sql#686]
I'm struggling to find a programmatic reproducer for this, but on the datafusion-sql-planner branch:
```
c.sql("SELECT * FROM large_table limit 5")
```
results in reading the entire dataset before filtering at the end, instead of reading from a single partition.

Less reproducible, but from the daily weather data:
```
res = c.sql("select * from weather limit 5")
io_layer = list(res.dask.layers.keys())[0]
partitions = len(list(res.dask.layers[io_layer].keys()))
partitions
```
```
445
```

If I understand layers correctly, my select w/ limit statement is reading all partitions in the dataset.

**Comment by sarahyurick** [comment]
When examining the HighLevelGraph with `df.dask`, it looks like the 445 being returned represents the number of partitions being specified when initially reading in the Dask DataFrame with `from_pandas` (or however the the data is being read), e.g. `df = dd.from_pandas(df, npartitions=445)`.

But you are correct that whenever we are using the `LIMIT` statement from `dask_sql/physical/rel/logical/limit.py`, we are returning `df.head(limit, npartitions=-1, compute=False)` and using all partitions. The way to reproduce this is with something like:

```
from dask_sql import Context
import dask.dataframe as dd
import numpy as np
import pandas as pd
import sqlite3

c = Context()
iris = pd.read_csv("iris.csv")
iris = dd.from_pandas(iris, npartitions=5)
c.create_table("iris", iris)

res = c.sql("SELECT * FROM iris LIMIT 3")
io_layer = list(res.dask.layers.keys())[2]
partitions = len(list(res.dask.layers[io_layer].keys()))
print("partitions:", partitions)
```

Prints `partitions: 6` (more generally, npartitions+1). I think we're also technically returning the `number of outputs` instead of `npartitions` with this, but I'm not quite sure what the difference is. Note that we specify 'io_layer = list(res.dask.layers.keys())[2]' since `head()` will be the 3rd (out of 5) layer of the HighLevelGraph.

When we change it to `df.head(limit, npartitions=1, compute=False)` in `limit.py`, I believe we get the desired behavior as the above code block prints 1 partition. We'll just have to be careful that `limit` < the size of the partition, and set `npartitions` to the smallest value possible.

(As an additional example, if we change it to `df.head(limit, npartitions=2, compute=False)` in `limit.py`, then the code block prints 3 partitions.)

Let me know if any of my logic looks off, but it looks like the entire fix can be done in `limit.py`.

**Comment by ayushdg** [comment]
You are right that `limit` statements now returns the whole dataframe via `npartitions=-1`. 

Another approach used in the main branch today is to return the computed `head` from the first partition and if the number of elements are lesser than the limit we've asked for, we read the whole dataset.

There are drawbacks to either approach. For more complex graphs with IO, joins etc. the second approach would compute the whole task graph twice (once for the getting the result for partition 1, and once more incase the limit>num_elements). 

We can switch back to using that approach if desired but there's likely going to be duplicated compute in some cases no matter which approach we take. 

cc: @charlesbluca

**Comment by charlesbluca** [comment]
Yeah for some additional context this change was made with https://github.com/dask-contrib/dask-sql/pull/529; the consensus was that (as @ayushdg said) there are some pros and cons to either method depending on input, and that it might make sense to be able to configure something like this in the future on a per-query basis.

**Comment by sarahyurick** [comment]
I see what you mean, although we can get the length of each partition with `partition_borders = df.map_partitions(lambda x: len(x)).divisions`, for example if the DataFrame has 150 rows and 5 partitions then `partition_borders = (0, 30, 60, 90, 120, 149)`. Would it be worth it to use that info to set the number of partitions in `head()`, e.g., `npartitions=1` for 0<limit<=30, `npartitions=2` for 30<limit<=60, etc.?

**Comment by charlesbluca** [comment]
> although we can also get the length of each partition with `partition_borders = df.map_partitions(lambda x: len(x)).divisions`

That's a great point that I hadn't considered! Knowing that, I think we can probably update our current method for computing LIMITs on main, which as @ayushdg has a chance of forcing 2 full graph computes:

https://github.com/dask-contrib/dask-sql/blob/3ac2f6df6f806e2c2fc5a9ff4b4bfbaa84d931cb/dask_sql/physical/rel/logical/limit.py#L53-L58

If we replace the assignment of `first_partition_length` with a call to compute `partition_borders`, then we will have all the partition lengths and guarantee that we can compute the limit with `df.head(...)`.

> Update: Nevermind, looks like that causes some tests in `test_join.py` and `test_select.py` to fail.

I currently have a small WIP that seems to be passing, I can open a PR to continue discussion there 🙂

**Comment by ayushdg** [comment]
> I currently have a small WIP that seems to be passing, I can open a PR to continue discussion there
It would be great to continue the discussion there. Partition_borders can be useful for this, but I have noticed some situations when reading from csv/parquet etc where the divisions are not known unless we set index/sort. In those cases we'd probably still need some duplicated compute.

**Comment by sarahyurick** [comment]
Sounds good! I'm also down for continuing the discussion there - I'm not sure why the failures were happening on my end, so it will be helpful to compare.
